### PR TITLE
Issue 3: Add timeline store and diff model for tracer events 

### PR DESCRIPTION
**Pull request by NicholasTc**
## Summary
- Add a new UI-agnostic timeline storage module at `Lib/idlelib/timeline_store.py`.
- Implement step/event APIs and deterministic diffing for locals snapshots:
  - `store_event(event)`
  - `get_step(i)` (0-based, raises `IndexError` if out of range)
  - `diff(prev, curr)` and `diff_steps(i, j)`
- Add tests at `Lib/idlelib/idle_test/test_timeline_store.py` for ordering, out-of-range behavior, diff correctness, and edge-case safety.

## Why
Issue #2 provides raw per-line trace events; this PR adds the model layer that converts those events into a stable, UI-ready timeline contract (`added`, `removed`, `changed`) for Issue #4.

## Behavior / Contract
- Default diff scope is `locals`.
- Diff result shape:
  - `added: {name: value}`
  - `removed: {name: value}`
  - `changed: {name: {"before": old, "after": new}}`
- Non-mapping or missing snapshot namespaces are handled gracefully (treated as empty snapshots).
- Unusual value comparison/repr cases are handled defensively to avoid crashes.

## Tests
- Added: `idlelib.idle_test.test_timeline_store`
- Re-validated with existing tracer test:
  - `idlelib.idle_test.test_timeline_tracer`

## Scope Notes
- No debugger/UI integration changes in this PR.
- No changes to `debugger.py` or timeline UI modules; this is storage/model only.

### CODE REVIEW

File: Lib/idlelib/timeline_store.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    except Exception as exc:
+        return f"<unreprable {type(value).__name__}: {exc!r}>"
+
+
+def _values_equal(left: Any, right: Any) -> bool:
+    """Compare two values without allowing unusual __eq__ to crash diffing."""
+    try:
+        return left == right
+    except Exception:
+        return _safe_repr(left) == _safe_repr(right)
```
Comment (evanedreo) [inline Lib/idlelib/timeline_store.py]: Nice defensive handling of unusual `__eq__` implementations, that prevents diffing from crashing.

File: Lib/idlelib/idle_test/test_timeline_store.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        store = TimelineStore()
+        store.store_event({"step": 1, "locals": None})
+        store.store_event({"step": 2, "locals": 123})
+
+        result = store.diff_steps(0, 1)
+        self.assertEqual(result["added"], {})
+        self.assertEqual(result["removed"], {})
+        self.assertEqual(result["changed"], {})
+
+    def test_diff_handles_unusual_values_without_crashing(self):
```
Comment (evanedreo) [inline Lib/idlelib/idle_test/test_timeline_store.py]: Good edge-case test that directly validates the acceptance criteria about handling broken

**Review by evanedreo — APPROVED** [review]
## Summary 

This PR introduces a new UI-agnostic `timeline_store` module that stores tracer events and provides deterministic diffing between snapshots (defaulting to `locals`). It defines `store_event`, `get_step`, `diff`, and `diff_steps`, and adds unit tests covering ordering, out-of-range access, diff correctness, and defensive handling of unusual `__eq__`/`__repr__` behavior.

---

## Correctness

* [x] Core functionality works as intended
* [x] Handles edge cases and error conditions appropriately

### Evidence / Notes

* Reviewed `test_timeline_store.py`, which covers:

  * Event ordering and retrieval via `get_step`
  * Out-of-range `get_step` raising `IndexError`
  * Correct detection of `added`, `removed`, and `changed`
  * Handling of missing or non-mapping `locals`
  * Defensive handling of objects with broken `__eq__` or `__repr__`
* The `diff` logic correctly computes set differences for keys and uses `_values_equal` to avoid crashes during equality checks.
* `_snapshot_from_event` safely treats non-mapping namespaces as empty dicts, satisfying the “graceful handling” acceptance criteria.
* Behavior aligns with the issue’s acceptance criteria (safe repr, no UI integration, correct diff semantics).

---

## Design

* [x] Exemplary - Clean, modular, easy to follow. Strong separation of concerns. Minimal complexity.

### Explain your choice

* The module is clearly UI-agnostic and isolated from debugger/UI concerns, matching the scope notes.
* Responsibilities are well separated:

  * `diff` handles pure snapshot comparison.
  * `_safe_repr` and `_values_equal` encapsulate defensive comparison logic.
  * `TimelineStore` focuses only on event storage and step-based access.
* Functions are short, single-purpose, and readable.
* The namespace abstraction (`namespace="locals"`) is a good design choice that keeps the store flexible for future extensions (e.g., globals).
* No obvious duplication or overly complex logic.

---

## Code Style

* [x] Exemplary - Consistent formatting, clear naming, helpful comments, follows style guide.

### Explain your choice

* [x] Variable/function/class names are descriptive and consistent
* [x] Formatting/indentation is consistent
* [x] Comments/docstrings clarify non-obvious logic
* [x] Code follows project/course style guidelines

Notes:

* Docstrings clearly describe return shapes and intent.
* Private helpers (`_safe_repr`, `_values_equal`, `_snapshot_from_event`) are well-named and scoped appropriately.
* Type hints are consistently applied and improve readability.
* The test cases are clean, focused, and descriptive.

---

## Inline Comments

Did you add any inline comments in the Files Changed tab?

* [x] Yes
* [ ] No

I added inline comments on defensive comparison logic and the test.
